protocol/src/packet.rs

```diff
@@ -0,0 +1,190 @@
+use std::io::{self, Read, Write};
+
+use flate2::read::ZlibDecoder;
+use flate2::write::ZlibEncoder;
+use flate2::Compression;
+use minecraft_protocol_derive::{Decoder, Encoder};
+
+use crate::decoder::{Decoder, DecoderReadExt};
+use crate::encoder::{Encoder, EncoderWriteExt};
+use crate::error::{DecodeError, EncodeError};
+
+#[derive(Debug, Clone)]
+pub struct Packet {
+    pub id: i32,
+    pub data: Vec<u8>,
+}
+
+impl Packet {
+    pub fn encode<W: Write>(
+        &self,
+        writer: &mut W,
+        compression_threshold: Option<i32>,
+    ) -> Result<(), EncodeError> {
+        let mut buf = Vec::new();
+        let packet = RawPacket {
+            id: self.id,
+            data: self.data.clone(),
```

Comment:
Can we avoid data cloning?

Reply:
Changed the method to take an owned `self`.